pkg/cloud/services/compute/instance.go

```diff
@@ -407,68 +386,143 @@ func (s *Service) getVolumeByName(name string) (*volumes.Volume, error) {
+	availabilityZone := instanceSpec.FailureDomain
+	if dataVolume.AvailabilityZone != "" {
+		availabilityZone = dataVolume.AvailabilityZone
 	}
-	record.Eventf(eventObject, "SuccessfulCreateVolume", "Created root volume; id=%s", volume.ID)
-	return volume, err
+
+	createOpts := volumes.CreateOpts{
+		Name:             volumeName(instanceSpec.Name, dataVolume.NameSuffix),
+		Description:      fmt.Sprintf("Block device for %s", instanceSpec.Name),
+		Size:             dataVolume.Size,
+		Multiattach:      false,
+		AvailabilityZone: availabilityZone,
+		VolumeType:       dataVolume.VolumeType,
+	}
+
+	return s.getOrCreateVolume(eventObject, createOpts)
 }
 
-// applyRootVolume sets a root volume if the root volume Size is not 0.
-func applyRootVolume(opts servers.CreateOptsBuilder, volume *volumes.Volume) servers.CreateOptsBuilder {
-	if volume == nil {
-		return opts
+func (s *Service) getBlockDevices(eventObject runtime.Object, instanceSpec *InstanceSpec, imageID string, timeout time.Duration, retryInterval time.Duration) ([]bootfromvolume.BlockDevice, error) {
+	blockDevices := []bootfromvolume.BlockDevice{}
+
+	if hasRootVolume(instanceSpec) {
+		rootVolume, err := s.getOrCreateRootVolume(eventObject, instanceSpec, imageID)
+		if err != nil {
+			return []bootfromvolume.BlockDevice{}, err
+		}
+		blockDevices = append(blockDevices, bootfromvolume.BlockDevice{
+			SourceType:          bootfromvolume.SourceVolume,
+			DestinationType:     bootfromvolume.DestinationVolume,
+			UUID:                rootVolume.ID,
+			BootIndex:           0,
+			DeleteOnTermination: true,
+		})
+	} else {
+		blockDevices = append(blockDevices, bootfromvolume.BlockDevice{
+			SourceType:          bootfromvolume.SourceImage,
+			DestinationType:     bootfromvolume.DestinationLocal,
+			UUID:                imageID,
+			BootIndex:           0,
+			DeleteOnTermination: true,
+		})
 	}
 
-	block := bootfromvolume.BlockDevice{
-		SourceType:          bootfromvolume.SourceVolume,
-		BootIndex:           0,
-		UUID:                volume.ID,
-		DeleteOnTermination: true,
-		DestinationType:     bootfromvolume.DestinationVolume,
+	if instanceSpec.BlockDevices != nil && len(instanceSpec.BlockDevices) > 0 {
+		for _, dataVolumeSpec := range instanceSpec.BlockDevices {
+			blockDevice, err := s.getOrCreateBlockDevice(eventObject, instanceSpec, dataVolumeSpec)
+			if err != nil {
+				return []bootfromvolume.BlockDevice{}, err
+			}
+			blockDevices = append(blockDevices, bootfromvolume.BlockDevice{
+				SourceType:          bootfromvolume.SourceVolume,
+				DestinationType:     bootfromvolume.DestinationVolume,
+				UUID:                blockDevice.ID,
+				BootIndex:           -1,
+				DeleteOnTermination: true,
+				// TODO(mkjpryor) uncomment once gophercloud patch is merged
+				// https://github.com/gophercloud/gophercloud/pull/2777
+				// Tag:                 dataVolumeSpec.Tag,
```

Comment:
Once https://github.com/gophercloud/gophercloud/pull/2789 is merged and a release was produced, I'll update vendoring here.